samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -110,17 +109,92 @@ public ZkKeyBuilder getKeyBuilder() {
+      zkClient.delete(ephemeralPath);
+      throw new SamzaException(String.format("Processor: %s is duplicate in the group. Registration failed.", processorId));
     }
+    return ephemeralPath;
+  }
+
+  /**
+   * Determines if a processor should be stopped.
+   * @param processor to check for stop condition in processors group.
+   * @return true if processor should be stopped. false otherwise.
+   */
+  private boolean shouldStopProcessor(final ProcessorNode processor) {
+    /**
+     * If there are multiple processors registered with same processorId,
+     * allow only the processor with lexicographically smallest zookeeperPath to live
+     * and all the remaining processors should be stopped.
+     *
+     * Two processors will not have smallest zookeeperPath because of sequentialId guarantees
+     * of zookeeper for ephemeral nodes.
+     */
+    List<ProcessorNode> processorNodes = getAllProcessorNodes();
+    LOG.debug("All existing processor nodes in zookeeper: {}.", processorNodes);
+    TreeSet<String> sortedProcessorPaths = processorNodes.stream()
+                                                         .filter(processorNode -> processorNode.getProcessorData().getProcessorId().equals(processor.getProcessorData().getProcessorId()))
+                                                         .map(ProcessorNode::getEphemeralPath)
+                                                         .collect(Collectors.toCollection(TreeSet::new));
+    return !sortedProcessorPaths.first().equals(processor.getEphemeralPath());
+  }
+
+  /**
+   * Find the number of processors having the provided processorId.
+   * @param processorId to use for equality check.
+   * @return the count of matched processors.
+   */
+  private long getProcessorCount(final String processorId) {
+    List<ProcessorNode> processorNodes = getAllProcessorNodes();
+    long processorCount = processorNodes.stream()
+                                        .filter(processorNode -> processorNode.processorData.getProcessorId().equals(processorId))
+                                        .count();
+    LOG.debug("Number of processors with id: {} = {}.", processorId, processorCount);
+    return processorCount;
+  }
+
+  /**
+   * Fetches all the ephemeral processor nodes of a standalone job from zookeeper.
+   * @return a list of {@link ProcessorNode}, where each ProcessorNode represents a registered stream processor.
+   */
+  private List<ProcessorNode> getAllProcessorNodes() {
+    List<String> processorZNodes = getSortedActiveProcessorsZnodes();
+    LOG.debug("Active ProcessorZNodes in zookeeper: {}.", processorZNodes);
+    return processorZNodes.stream()
+                          .map(processorZNode -> {
+                              String ephemeralProcessorPath = String.format("%s/%s", keyBuilder.getProcessorsPath(), processorZNode);
+                              String data = zkClient.readData(ephemeralProcessorPath);
```

Comment:
Is it possible to use `ZkUtils#readProcessorData` instead of using zkclient directly ?
Metrics is lost here since we are not using wrapper methods for zkClient. There are other places in this PR where we are doing it this way. It may also be the same in the existing code. I think we should create wrapper methods for every zkclient api that we use for the sake of metrics.